lib/backends/sdc/containers.js

```diff
@@ -981,130 +999,387 @@ function renameLinks(opts, newName, callback) {
+    var nameMatches = [];
+    var uuidMatches = [];
+
+    constraint[PACKAGE_SELECTION_LABEL] = specifiedPackage;
+
+    if (pkgs.length === 1) {
+        // The lookup we did against PAPI returned only 1 package. That's our
+        // target.
+        foundUuid = pkgs[0].uuid;
+    } else if (pkgs.length > 1) {
+        // This should mean that there's both a name and uuid that match, we'll
+        // use the one where the name matches.
+        pkgs.forEach(function _onEachPkg(pkg) {
+            if (pkg.name === specifiedPackage) {
+                nameMatches.push(pkg.name);
+            } else if (specifiedPackage.match(/^[0-9a-f]{8}$/)
+                && (pkg.uuid.substr(0, 8) === specifiedPackage)) {
+
+                uuidMatches.push(pkg.uuid);
+            } else if (pkg.uuid === specifiedPackage) {
+                uuidMatches.push(pkg.uuid);
             }
+        });
 
-            log.debug({pkg: candidate}, 'selected package for VM');
-            callback(null, candidate.uuid);
+        // We're trying to uniquely identify a package, so only allow one match
+        if (nameMatches.length === 1) {
+            foundUuid = nameMatches[0];
+        } else if (uuidMatches.length === 1) {
+            foundUuid = uuidMatches[0];
         }
-    );
+
+        // If there are more than 1 match of either, or if there are no matches,
+        // we can't determine the package to use and will fail.
+    }
+
+    if (foundUuid) {
+        log.debug({foundPackage: foundUuid}, 'found package');
+        callback(null, foundUuid);
+        return;
+    }
+
+    log.error({specifiedPackage: specifiedPackage, pkgs: pkgs},
+        'unable to find specified package');
+    callback(new errors.NoSufficientPackageError(constraint));
 }
 
+/*
+ * This returns a filter string for querying PAPI to get the set of potential
+ * packages.
+ *
+ * When doing a lookup the precedence is:
+ *
+ *  - uuid
+ *
+ *      If the argument is a UUID, we'll only match UUID
+ *
+ *  - name
+ *
+ *      If the argument is /^[0-9a-f]{8}$/ and matches both a uuid and a name,
+ *      the package with the name that matches is used. If the argument does not
+ *      match the short-UUID pattern, and is not a UUID, it's only looked up
+ *      against package names.
+ *
+ *  - short-UUID
+ *
+ *      If the argument is /^[0-9a-f]{8}$/ and does not match a name, it will
+ *      be looked up against the first 8 characters of the available package
+ *      UUIDs.
+ *
+ */
+function buildPackageFilter(opts) {
+    assert.object(opts, 'opts');
+    assert.string(opts.packagePrefix, 'opts.packagePrefix');
+    assert.uuid(opts.ownerUuid, 'opts.ownerUuid');
+    assert.optionalString(opts.specifiedPackage, 'opts.specifiedPackage');
+
+    var filter;
+
+    if (opts.specifiedPackage) {
+        if (common.isUUID(opts.specifiedPackage)) {
+            // if specifiedPackage is a UUID, we'll just look that up directly.
+            filter = '(&(active=true)'
```

Comment:
@jclulow Absolutely, what I wasn't 100% sure of is not about translating the `!` operator, but more if `ldap-filter` was doing anything special with the `*` pattern. In other words, what I meant is: "please check that using ldap-filter's filters does a direct translation and what we expect, because I haven't checked that myself, so I'm not 100% sure".